### PR TITLE
Modify registry data to adjust Windows service configuration

### PR DESCRIPTION
**Pull request by RebeccaMahany**
Applies the changes from https://github.com/kolide/launcher/pull/1142 to existing installations.

### CODE REVIEW

File: cmd/launcher/svc_windows.go [hunk 1]
```diff
@@ -190,3 +194,43 @@ func (w *winSvc) Execute(args []string, r <-chan svc.ChangeRequest, changes chan
...
+
+	// DependOnService is type REG_MULTI_SZ, i.e. a list of strings
+	dependOnServiceName = `DependOnService`
+	dependOnService     = `Dnscache`
+)
+
+func editRegistry(logger log.Logger) {
+	// Get launcher service key
+	// TODO: ideally we'd pull the identifier from somewhere legitimate
+	launcherServiceKeyName := fmt.Sprintf(launcherServiceRegistryKeyNameFormat, "KolideK2")
```
Comment (directionless) [inline cmd/launcher/svc_windows.go:213]: Can we pull it from the config file?

**Review by directionless — COMMENTED** [review]
This looks pretty simple. What, if any, do you think the risks are?

File: cmd/launcher/svc_windows.go (hunk 1 above)
Comment (RebeccaMahany) [reply cmd/launcher/svc_windows.go:213]: My thought was that we'd extract it from the server hostname in the options, but hadn't spent the time chasing that down yet to make sure that was correct. If yes I can definitely do that, the options are available here.

**Comment by RebeccaMahany** [comment]
> This looks pretty simple. What, if any, do you think the risks are?

In test, I didn't run into any issues -- it seems to be fairly accepted to use this API to edit your own application's data. To avoid unnecessary writes, I would update to check the values first and only re-set them if they aren't already correct, instead of writing every time.

**Comment by directionless** [comment]
> > This looks pretty simple. What, if any, do you think the risks are?
> 
> In test, I didn't run into any issues -- it seems to be fairly accepted to use this API to edit your own application's data. To avoid unnecessary writes, I would update to check the values first and only re-set them if they aren't already correct, instead of writing every time.

If this is accepted, then it's probably consistent across windows versions? We should probably do something to check that there's any kind of expected value, before overwriting it. (eg: if it's not set, then the system is different, and 🤷 )

Any chance we'd have a partial write that might break service restart?

**Comment by RebeccaMahany** [comment]
> If this is accepted, then it's probably consistent across windows versions? We should probably do something to check that there's any kind of expected value, before overwriting it. (eg: if it's not set, then the system is different, and 🤷 )

I added a check yesterday to see if the values are already set to what we expect, so that we don't perform unnecessary writes in that case. If we fail to pull the values, then we don't perform the write. I think that should cover this? Let me know if I'm missing something.

> Any chance we'd have a partial write that might break service restart?

I don't think we're in too much danger with flipping the delayed autostart bool, but it might be possible that we could write bad/incomplete data to the list of services that launcher depends on? I'll try out writing a nonexistent service to the array and see if it affects whether launcher can start itself after reboot.

For reference, it's ultimately using [this API](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regsetvalueexw) to set the values.

**Comment by RebeccaMahany** [comment]
> I'll try out writing a nonexistent service to the array and see if it affects whether launcher can start itself after reboot.

So yep that will absolutely stop launcher from being able to start itself after reboot. I feel like the chances of us getting it into this state are low, but maybe to be safest, we should not set `DependsOn` here -- that seems fine given that we still have the DNS checks on launcher first starting anyway. Thoughts?

**Review by James-Pickett — APPROVED** [review]

File: cmd/launcher/svc_windows.go [hunk 2]
```diff
@@ -190,3 +196,74 @@ func (w *winSvc) Execute(args []string, r <-chan svc.ChangeRequest, changes chan
 		}
 	}
 }
+
+const (
```
Comment (directionless) [inline cmd/launcher/svc_windows.go:200]: I'd toss all this into `cmd/launcher/svc_checker_windows.go` or something

**Review by directionless — APPROVED** [review]
I'd take it, but I think the flow is very hard to follow, especially around the service dependancy.

I suspect splitting some things into other functions and using early returns would help. At least, that's my usual go-to.

File: cmd/launcher/svc_windows.go (hunk 2 above)
Comment (RebeccaMahany) [reply cmd/launcher/svc_windows.go:200]: Will do!

**Review by seejdev — APPROVED** [review]
LGTM

File: cmd/launcher/svc_config_windows.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		// we don't want to wipe any current data from the list.
+		return
+	}
+
+	// Check whether the service configuration already includes Dnscache in the list of services it depends on
+	foundDnscacheInList := false
+	for _, service := range serviceList {
+		if service == dnscacheService {
+			foundDnscacheInList = true
+			break
```
Comment (directionless) [inline cmd/launcher/svc_config_windows.go:100]: You can early return the whole function, right?

**Review by directionless — DISMISSED** [review]

File: cmd/launcher/svc_config_windows.go (hunk 3 above)
Comment (RebeccaMahany) [reply cmd/launcher/svc_config_windows.go]: Yes, thank you! updated

**Review by directionless — APPROVED** [review]

File: cmd/launcher/svc_config_windows.go (hunk 3 above)
Comment (directionless) [reply cmd/launcher/svc_config_windows.go]: I <3 early returns
